### PR TITLE
[Admin][Order] Add order notes

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

**Comment by pjedrzejewski** [comment]
@Arminek There is a field ``additionalInformation`` on the order already...

**Comment by pjedrzejewski** [comment]
@Arminek We could rename it to notes indeed.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_notes.html.twig [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{% if order.notes is defined and order.notes is not null %}
+    <h4 class="ui top attached styled header">
+        {{ 'sylius.ui.notes'|trans }}
+    </h4>
+    <div class="ui attached segment" id="sylius-order-notes">
+        {{ order.notes }}
+    </div>
+{% endif %}
```
Comment (tuka217) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_notes.html.twig]: Missing blank line

File: src/Sylius/Behat/Page/Shop/Checkout/SummaryPage.php [hunk 2]
```diff
@@ -134,6 +142,7 @@ protected function getDefinedElements()
             'items_table' => '#items table',
             'product_row' => 'tbody tr:contains("%name%")',
             'order_total' => 'td:contains("Total")',
+            'extra_notes' =>'#sylius_checkout_summary_notes',
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Shop/Checkout/SummaryPage.php]: Why not just ``notes``? Add I would keep alphabetical order ;)

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_notes.html.twig [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{% if order.notes is defined and order.notes is not null %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_notes.html.twig]: I think ``order.notes`` is always defined, as it's ``Order``'s property, so ``is not null`` should be sufficient, wdyt?

File: src/Sylius/Component/Order/Model/Order.php [hunk 4]
```diff
@@ -153,6 +153,22 @@ public function setNumber($number)
     /**
      * {@inheritdoc}
      */
+    public function getNotes()
```
Comment (Zales0123) [inline src/Sylius/Component/Order/Model/Order.php]: Don't we need any spec updates for that?

File: features/order/placing_an_order/adding_notes_for_customer_service.feature [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    In order to provide some extra information to order
+    As a Customer
+    I want to be able to provide a note for customer service
+
+    Background:
+        Given the store operates on a single channel in "France"
+        And the store has a product "PHP T-Shirt" priced at "$19.99"
+        And the store ships everywhere for free
+        And the store allows paying offline
+        And the store has customer service account identified by "sylius@example.com"
```
Comment (michalmarcinkowski) [inline features/order/placing_an_order/adding_notes_for_customer_service.feature]: `customer service account` let's not introduce another term for administrator user :) Should be renamed to admin.

Comment (michalmarcinkowski) [inline features/order/placing_an_order/adding_notes_for_customer_service.feature]: `And there is an administrator`

File: src/Sylius/Behat/Context/Transform/CustomerContext.php [hunk 6]
```diff
@@ -43,6 +43,7 @@ public function __construct(CustomerRepositoryInterface $customerRepository, Fac
 
     /**
      * @Transform :customer
+     * @Transform /^made by "[^"]+"$/
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Transform/CustomerContext.php]: This should only `getCustomerByEmail` and never create a customer. Add second method to this context.

File: src/Sylius/Behat/Context/Transform/OrderContext.php [hunk 7]
```diff
@@ -47,6 +58,21 @@ public function getOrderByNumber($orderNumber)
...
+    public function getOrderByCustomer($email)
+    {
+        $customer = $this->customerRepository->findOneBy(['email' => $email]);
+        Assert::notNull($customer, sprintf('Cannot find customer with email %s.', $email));
+
+        $order = $this->orderRepository->findByCustomer($customer);
+        $order = end($order);
+        Assert::isInstanceOf($order, OrderInterface::class, sprintf('Cannot find order for %s.', $customer->getFullName()));
+
+        return $order;
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Transform/OrderContext.php]:
```
$order = $this->orderRepository->findByCustomer($customer);
Assert::notEmpty($order);

return end($order);
```

File: src/Sylius/Behat/Context/Transform/OrderContext.php [hunk 8]
```diff
@@ -47,6 +58,21 @@ public function getOrderByNumber($orderNumber)
     }
 
     /**
+     * @Transform /^this order made by "([^"]+)"$/
+     */
+    public function getOrderByCustomer($email)
+    {
+        $customer = $this->customerRepository->findOneBy(['email' => $email]);
+        Assert::notNull($customer, sprintf('Cannot find customer with email %s.', $email));
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Transform/OrderContext.php]: If we use this transformer then why we have https://github.com/Sylius/Sylius/pull/5319/files#r68388816?

File: src/Sylius/Behat/Page/Admin/Order/ShowPage.php [hunk 9]
```diff
@@ -340,6 +341,17 @@ public function getRouteName()
     /**
      * {@inheritdoc}
      */
+    public function hasNote($note)
+    {
+        $orderNotesElement = $this->getElement('order_notes');
+        Assert::notNull($orderNotesElement, 'Cannot find order notes');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Page/Admin/Order/ShowPage.php]: IIRC `getElement('order_notes')` throws exception when the element is not found.

File: src/Sylius/Behat/Context/Setup/UserContext.php [hunk 10]
```diff
@@ -133,6 +133,28 @@ public function accountWasDeleted($email)
...
+    {
+        $user = $this->sharedStorage->get('user');
+
+        $this->userRepository->remove($user);
+    }
+
+    /**
+     * @Given there is an administrator identified by :email
+     */
+    public function theStoreHasCustomerServiceAccountIdentifiedBy($email)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: Please change customer service to administrator in shared storage and method name too.

**Pull request merged**